### PR TITLE
#251 EA-1986 If-Match check is fixed and ETag is updated

### PR DESCRIPTION
**Pull request by rgraf**

### CODE REVIEW

File: annotation-utils/src/main/java/eu/europeana/annotation/config/AnnotationConfiguration.java [hunk 1]
```diff
@@ -67,6 +69,8 @@
     public static final String VALIDATION_ADMIN_SECRET_KEY = "adminsecretkey";
 
     public static final String API_KEY_CACHING_TIME = "annotation.apikey.caching.time";
+    
+    public static final String ETAG_FORMAT = "application/json";
```
Comment (gsergiu) [inline annotation-utils/src/main/java/eu/europeana/annotation/config/AnnotationConfiguration.java]: please remove this constant and use the existing constant WebAnnotationFields.FORMAT_JSONLD

File: annotation-utils/src/main/java/eu/europeana/annotation/config/AnnotationConfiguration.java [hunk 2]
```diff
@@ -75,4 +79,11 @@
      */
     public Set<String> getAcceptedLicenceses();
 
+	/**
+	 * uses annotation.apiVersion property
+	 */
+	public String getAnnotationApiVersion();
+	
+	public String getEtagFormat();
```
Comment (gsergiu) [inline annotation-utils/src/main/java/eu/europeana/annotation/config/AnnotationConfiguration.java]: please remove this method, it must not be in configurations

File: annotation-utils/src/main/java/eu/europeana/annotation/config/AnnotationConfigurationImpl.java [hunk 3]
```diff
@@ -92,4 +92,15 @@ public String getMetisBaseUrl() {
...
 		return getAnnotationProperties().getProperty(key);
 	}
+
+    @Override
+    public String getAnnotationApiVersion() {
+    	return getAnnotationProperties().getProperty(API_VERSION);
+    }
+
+    @Override
+    public String getEtagFormat() {
```
Comment (gsergiu) [inline annotation-utils/src/main/java/eu/europeana/annotation/config/AnnotationConfigurationImpl.java]: remove

File: annotation-web/src/main/java/eu/europeana/annotation/web/service/controller/jsonld/BaseJsonldRest.java [hunk 4]
```diff
@@ -322,15 +326,15 @@ protected Agent buildAgent(String id) {
 	    JsonLd annotationLd = new AnnotationLdSerializer(annotation);
 	    String jsonLd = annotationLd.toString(4);
 
-	    int etag;
+	    String eTag;
 	    if (annotation.getLastUpdate() != null)
```
Comment (gsergiu) [inline annotation-web/src/main/java/eu/europeana/annotation/web/service/controller/jsonld/BaseJsonldRest.java]: please remove the use of lastUpdated, this was replaced now by generated

File: annotation-web/src/main/java/eu/europeana/annotation/web/service/controller/jsonld/BaseJsonldRest.java [hunk 5]
```diff
@@ -504,6 +513,50 @@ private boolean hasAdminRights(Authentication authentication) {
...
     }
 
+    /**
+     * This method compares If-Match header with the current etag value.
+     * 
+     * @param etag    The current etag value
+     * @param request The request containing If-Match header
+     * @throws HttpException
+     */
+    public void checkIfMatchHeader(String etag, HttpServletRequest request) throws HttpException {
```
Comment (gsergiu) [inline annotation-web/src/main/java/eu/europeana/annotation/web/service/controller/jsonld/BaseJsonldRest.java]: the same methos exists in Entity API, please move implementation to API commons and also remove the method from Entity API

File: annotation-web/src/main/java/eu/europeana/annotation/web/service/controller/jsonld/BaseJsonldRest.java [hunk 6]
```diff
@@ -504,6 +513,50 @@ private boolean hasAdminRights(Authentication authentication) {
...
+		}
+    }
+    
+    /**
+     * This method generates etag for response header.
+     * 
+     * @param modifiedDate The date of the last modification of annotaiton
+     * @return etag value
+     */
+    public String generateETag(Date modifiedDate) {
```
Comment (gsergiu) [inline annotation-web/src/main/java/eu/europeana/annotation/web/service/controller/jsonld/BaseJsonldRest.java]: Please move this method to api commons and add the format and version parameters

File: annotation-web/src/main/java/eu/europeana/annotation/web/service/controller/jsonld/BaseJsonldRest.java [hunk 7]
```diff
@@ -504,6 +513,50 @@ private boolean hasAdminRights(Authentication authentication) {
...
+    	// add timestamp
+		Integer hashCode;
+		hashCode = modifiedDate.hashCode();
+		
+		// add ETag format 
+		hashCode += '|'+getConfiguration().getEtagFormat().hashCode();
```
Comment (gsergiu) [inline annotation-web/src/main/java/eu/europeana/annotation/web/service/controller/jsonld/BaseJsonldRest.java]: response format and api version must be provided as parameters

File: annotation-web/src/main/java/eu/europeana/annotation/web/service/controller/jsonld/BaseJsonldRest.java [hunk 8]
```diff
@@ -504,6 +513,50 @@ private boolean hasAdminRights(Authentication authentication) {
...
+		
+		// add the hascode of the api version
+		String apiVersion = getConfiguration().getAnnotationApiVersion();
+		if (apiVersion != null)
+		    hashCode += '|'+apiVersion.hashCode();
+	
+		return hashCode.toString();
```
Comment (gsergiu) [inline annotation-web/src/main/java/eu/europeana/annotation/web/service/controller/jsonld/BaseJsonldRest.java]: the implementation of this method should be simply: (timestamp+format+version).hasCode()

File: annotation-utils/src/main/java/eu/europeana/annotation/config/AnnotationConfiguration.java (hunk 1 above)
Comment (rgraf) [reply annotation-utils/src/main/java/eu/europeana/annotation/config/AnnotationConfiguration.java]: constant replaced

File: annotation-utils/src/main/java/eu/europeana/annotation/config/AnnotationConfiguration.java (hunk 2 above)
Comment (rgraf) [reply annotation-utils/src/main/java/eu/europeana/annotation/config/AnnotationConfiguration.java]: method removed

File: annotation-utils/src/main/java/eu/europeana/annotation/config/AnnotationConfigurationImpl.java (hunk 3 above)
Comment (rgraf) [reply annotation-utils/src/main/java/eu/europeana/annotation/config/AnnotationConfigurationImpl.java]: method removed

File: annotation-web/src/main/java/eu/europeana/annotation/web/service/controller/jsonld/BaseJsonldRest.java (hunk 4 above)
Comment (rgraf) [reply annotation-web/src/main/java/eu/europeana/annotation/web/service/controller/jsonld/BaseJsonldRest.java]: replaced by getGenerated

File: annotation-web/src/main/java/eu/europeana/annotation/web/service/controller/jsonld/BaseJsonldRest.java (hunk 5 above)
Comment (rgraf) [reply annotation-web/src/main/java/eu/europeana/annotation/web/service/controller/jsonld/BaseJsonldRest.java]: moved to commons-api

File: annotation-web/src/main/java/eu/europeana/annotation/web/service/controller/jsonld/BaseJsonldRest.java (hunk 6 above)
Comment (rgraf) [reply annotation-web/src/main/java/eu/europeana/annotation/web/service/controller/jsonld/BaseJsonldRest.java]: moved to commons-api

File: annotation-web/src/main/java/eu/europeana/annotation/web/service/controller/jsonld/BaseJsonldRest.java (hunk 7 above)
Comment (rgraf) [reply annotation-web/src/main/java/eu/europeana/annotation/web/service/controller/jsonld/BaseJsonldRest.java]: added format and version parameters

File: annotation-web/src/main/java/eu/europeana/annotation/web/service/controller/jsonld/BaseJsonldRest.java (hunk 8 above)
Comment (rgraf) [reply annotation-web/src/main/java/eu/europeana/annotation/web/service/controller/jsonld/BaseJsonldRest.java]: change method according to suggestion

Comment (gsergiu) [reply annotation-web/src/main/java/eu/europeana/annotation/web/service/controller/jsonld/BaseJsonldRest.java]: the method was moved now to api-commons it must be removed from this class

File: annotation-web/src/main/java/eu/europeana/annotation/web/service/controller/jsonld/BaseJsonldRest.java (hunk 7 above)
Comment (gsergiu) [reply annotation-web/src/main/java/eu/europeana/annotation/web/service/controller/jsonld/BaseJsonldRest.java]: the method was moved now to api-commons it must be removed from this class

File: annotation-web/src/main/java/eu/europeana/annotation/web/service/controller/jsonld/BaseJsonldRest.java (hunk 6 above)
Comment (gsergiu) [reply annotation-web/src/main/java/eu/europeana/annotation/web/service/controller/jsonld/BaseJsonldRest.java]: must remove the method from this class after refactoring

File: annotation-web/src/main/java/eu/europeana/annotation/web/service/controller/jsonld/BaseJsonldRest.java (hunk 5 above)
Comment (gsergiu) [reply annotation-web/src/main/java/eu/europeana/annotation/web/service/controller/jsonld/BaseJsonldRest.java]: the method must be removed now from this class

**Pull request merged**